### PR TITLE
*: support distributed aggregate function count

### PR DESCRIPTION
**Pull request by shenli**
I will support sum, avg, max, min and group_concat soon in this PR.
@coocood

### CODE REVIEW

File: executor/aggregate_xapi.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+			return false, nil
+		}
+	} else {
+		// If Src is nil, only one row should be returned.
+		if e.executed {
+			return false, nil
+		}
+	}
+	e.executed = true
+	groupKey := row.Data[0].GetBytes()
```
Comment (coocood) [inline executor/aggregate_xapi.go]: groupKey may not be `row.Data[0]`.

File: executor/builder.go [hunk 2]
```diff
@@ -426,7 +430,71 @@ func (b *executorBuilder) buildAggregate(v *plan.Aggregate) Executor {
...
+	for i, agg := range v.AggFuncs {
+		ft := types.NewFieldType(mysql.TypeLonglong)
+		ft.Flen = 21
+		ft.Charset = charset.CharsetBin
+		ft.Collate = charset.CollationBin
+		fields[2*i+1] = ft
+		fields[2*i+2] = agg.GetType()
+	}
+	fmt.Println("AggFields", len(fields))
+	xSrc.AddAggregate(pbAggFuncs, pbByItems, fields)
```
Comment (coocood) [inline executor/builder.go]: Move the logic to `AddAggregate` implementation and make `AddAggregate` return a bool value that indicates if aggregation can be pushed may be better.

Comment (shenli) [inline executor/builder.go]: If so, we need to implement this logic in multiple XExecutors including XTableScanExec, XIndexScanExec.

File: executor/builder.go [hunk 3]
```diff
@@ -426,7 +429,71 @@ func (b *executorBuilder) buildAggregate(v *plan.Aggregate) Executor {
...
+	for _, item := range v.GroupByItems {
+		pbByItem := b.groupByItemToPB(client, item, xSrc.GetTableName())
+		if pbByItem == nil {
+			return e
+		}
+		pbByItems = append(pbByItems, pbByItem)
+	}
+	// compose aggregate info
+	// We should infer fields type.
+	// Each agg item will be splited into two datum: count and value
```
Comment (coocood) [inline executor/builder.go]: splitted
two datums

**Comment by shenli** [comment]
@coocood @zimulala @hanfei19910905 @BusyJay PTAL
I will add some test cases in this PR.

File: store/localstore/local_aggregate.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	ds := make([]types.Datum, 2)
+	ds[0] = types.NewUintDatum(item.count)
+	ds[1] = item.value
+	return ds
+}
+
+var singleGroupKey = []byte("SingleGroup")
+
+// getContext gets aggregate evaluation context for the current group.
```
Comment (zimulala) [inline store/localstore/local_aggregate.go]: `getContext` -> `getAggItem` ?

File: store/localstore/local_aggregate.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+func (n *aggregateFuncExpr) toDatums() []types.Datum {
+	switch n.expr.GetTp() {
+	case tipb.ExprType_Count:
+		return n.getCountValue()
+	}
+	// TODO: return error
+	return nil
+}
+
+// Convert count and value to datum list
```
Comment (zimulala) [inline store/localstore/local_aggregate.go]: This comment need update ? 
Add . at the end of a line.

**Comment by shenli** [comment]
PTAL @zimulala

File: executor/aggregate_xapi.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+			hasMore, err := e.innerNext()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if !hasMore {
+				break
+			}
+		}
+		e.executed = true
+		if len(e.groups) == 0 {
```
Comment (coocood) [inline executor/aggregate_xapi.go]: Seems like this branch will never be executed, as each local region returns single groupKey if no group by items.

Comment (coocood) [inline executor/aggregate_xapi.go]: If there is 'group by' items, and the table is empty, we should get an empty result, so we can't add `singleGroup` in this case.

File: store/localstore/local_region.go [hunk 7]
```diff
@@ -140,6 +155,44 @@ func (rs *localRegion) getRowsFromSelectReq(ctx *selectContext) ([]*tipb.Row, er
...
+ * Convert aggregate partial result to rows.
+ * Data layout example:
+ *	SQL:	select count(c1), sum(c2) from t;
+ *	Aggs:	count(c1), sum(c2)
+ *	Rows:	groupKey1, count1, value1, count2, value2
+ *		groupKey2, count1, value1, count2, value2
+ */
+func (rs *localRegion) getRowsFromAgg(ctx *selectContext) ([]*tipb.Row, error) {
+	if len(ctx.groupKeys) == 0 {
+		ctx.groupKeys = append(ctx.groupKeys, singleGroup)
```
Comment (coocood) [inline store/localstore/local_region.go]: We should add singleGroup only if no group by items provided.

File: executor/aggregate_xapi.go (hunk 6 above)
Comment (shenli) [inline executor/aggregate_xapi.go]: Yes, you are right.

**Comment by shenli** [comment]
PTAL @coocood

File: store/localstore/local_aggregate.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	n.contextPerGroupMap = nil
+}
+
+// Update is used for update aggregate context.
+func (n *aggregateFuncExpr) update(ctx *selectContext, args []types.Datum) error {
+	switch n.expr.GetTp() {
+	case tipb.ExprType_Count:
+		return n.updateCount(ctx, args)
+	}
+	return errors.Errorf("Unknown AggExpr: %v", n.expr.GetTp())
```
Comment (zimulala) [inline store/localstore/local_aggregate.go]: Use terror.

File: store/localstore/local_region.go [hunk 9]
```diff
@@ -140,6 +155,41 @@ func (rs *localRegion) getRowsFromSelectReq(ctx *selectContext) ([]*tipb.Row, er
...
+	for _, gk := range ctx.groupKeys {
+		row := new(tipb.Row)
+		// Each aggregate partial result will be converted to two datum.
+		rowData := make([]types.Datum, 1+2*len(ctx.aggregates))
+		// The first column is group key.
+		rowData[0] = types.NewBytesDatum(gk)
+		for i, agg := range ctx.aggregates {
+			agg.currentGroup = gk
+			ds := agg.toDatums()
+			rowData[2*i+1] = ds[0]
```
Comment (coocood) [inline store/localstore/local_region.go]: ds can be nil

File: store/localstore/local_aggregate.go (hunk 8 above)
Comment (shenli) [inline store/localstore/local_aggregate.go]: This is mocked tikv region, so we do not use terror. And this should not happend.

File: store/localstore/local_region.go (hunk 9 above)
Comment (shenli) [inline store/localstore/local_region.go]: PTAL @coocood

File: store/localstore/local_region.go [hunk 10]
```diff
@@ -297,41 +350,58 @@ func (rs *localRegion) getRowByHandle(ctx *selectContext, handle int64) (*tipb.R
...
 			colID := col.GetColumnId()
 			if ctx.whereColumns[colID] != nil {
 				// The column is saved in evaluator, use it directly.
 				datum := ctx.eval.Row[colID]
-				row.Data, err = codec.EncodeValue(row.Data, datum)
-				if err != nil {
-					return nil, errors.Trace(err)
+				value := []byte{}
+				var err1 error
+				colVal, err1 = codec.EncodeValue(value, datum)
```
Comment (zimulala) [inline store/localstore/local_region.go]: Here `EncodeValue` can be used like `codec.EncodeValue(nil, datum)`.

File: store/localstore/local_aggregate.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	vals := make([]types.Datum, 0, len(items))
+	for _, item := range items {
+		v, err := ctx.eval.Eval(item.Expr)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		vals = append(vals, v)
+	}
+	bs, err := codec.EncodeValue([]byte{}, vals...)
```
Comment (zimulala) [inline store/localstore/local_aggregate.go]: The same as here.
`codec.EncodeValue(nil, vals...)`

Comment (shenli) [inline store/localstore/local_aggregate.go]: PTAL @coocood @zimulala

File: executor/aggregate_xapi_test.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
+// Copyright 2015 PingCAP, Inc.
```
Comment (zimulala) [inline executor/aggregate_xapi_test.go]: 2015 -> 2016

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
PTAL @zimulala

**Comment by zimulala** [comment]
LGTM

**Pull request merged**